torch/csrc/jit/passes/onnx/function_extraction.cpp

```diff
@@ -0,0 +1,1098 @@
+  }
+  do {
+    child = child->parent();
+    if (parent == child) {
+      return true;
+    }
+  } while (IsValidScope(child));
+  return false;
+}
+
+c10::optional<ScopePtr> FunctionExtractor::FindCommonAncestor(
+    ScopePtr a,
+    ScopePtr b) {
+  if (!IsValidScope(a) || !IsValidScope(b)) {
+    return c10::nullopt;
+  }
+
+  auto diff =
+      static_cast<int64_t>(a->getDepth()) - static_cast<int64_t>(b->getDepth());
+  if (diff != 0) {
+    auto deeper_scope = diff > 0 ? a : b;
+    auto other_scope = diff > 0 ? b : a;
+    while (diff > 0) {
+      deeper_scope = deeper_scope->parent();
+      diff--;
+    }
+    a = deeper_scope;
+    b = other_scope;
+  }
+
+  while (IsValidScope(a) && IsValidScope(b)) {
+    if (a == b) {
+      return a;
+    } else {
+      a = a->parent();
+      b = b->parent();
+    }
+  }
+
+  return c10::nullopt;
+}
+
+c10::optional<ScopePtr> FunctionExtractor::FindCommonAncestor(
+    const scope_list& scopes) {
+  if (scopes.size() == 0) {
+    return c10::nullopt;
+  }
+
+  c10::optional<ScopePtr> common_ancestor = scopes.at(0);
+  for (auto scope : scopes) {
+    common_ancestor = FindCommonAncestor(common_ancestor.value(), scope);
+    if (!common_ancestor.has_value()) {
+      return c10::nullopt;
+    }
+  }
+
+  return common_ancestor;
+}
+
+c10::optional<ScopePtr> FunctionExtractor::InferScope(Node* n) {
+  // The scope of node n is assigned based on the following rules.
+  // 1. If all uses of outputs of n belongs to the same scope,
+  //    assign that scope, otherwise
+  // 2. If all nodes of inputs of n belongs to the same scope,
+  //    assign that scope, otherwise
+  // 3. Find common ancestor of the scopes of uses of outputs of n,
+  //    and the scopes of nodes of inputs of n.
+  scope_list input_scopes;
+  scope_list output_scopes;
+  for (auto input : n->inputs()) {
+    input_scopes.emplace_back(input->node()->scope());
+  }
+  for (auto output : n->outputs()) {
+    for (auto use : output->uses()) {
+      output_scopes.emplace_back(use.user->scope());
+    }
+  }
+  if (output_scopes.size() > 0 &&
+      std::all_of(
+          output_scopes.begin(),
+          output_scopes.end(),
+          [&output_scopes](ScopePtr scope) -> bool {
+            return IsValidScope(scope) && scope == output_scopes.at(0);
+          })) {
+    return output_scopes.at(0);
+  } else if (
+      input_scopes.size() > 0 &&
+      std::all_of(
+          input_scopes.begin(),
+          input_scopes.end(),
+          [&input_scopes](ScopePtr scope) -> bool {
+            return IsValidScope(scope) && scope == input_scopes.at(0);
+          })) {
+    return input_scopes.at(0);
+  } else {
+    scope_list scopes;
+    std::copy_if(
+        input_scopes.begin(), input_scopes.end(), scopes.begin(), IsValidScope);
+    std::copy_if(
+        output_scopes.begin(),
+        output_scopes.end(),
+        scopes.begin(),
+        IsValidScope);
+
+    if (scopes.size() > 0) {
+      // find common ancestor.
```

Comment:
This comment is not adding any value. It just repeats what is obvious from the code immediately below it.